### PR TITLE
Basic Feature Gates

### PR DESCRIPTION
**Pull request by jojolepro**
## Description

This adds optional feature gates which can be used to cut compilation time down.
More will be added in the future (those that require more work).

## Additions

- Optional feature gates which will reduce compilation times when used.

## PR Checklist

By placing an x in the boxes I certify that I have:

- [X] Ran `cargo test --all` locally if this modified any rs files.
- [X] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- N/A Updated the content of the book if this PR would make the book outdated.
- [X] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- N/A Added unit tests for new APIs if any were added in this PR.
- [X] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

**Comment by khionu** [comment]
Could UI be added to this? I would say it's of the same priority as audio

**Comment by minecrawler** [comment]
I would even add the renderer and input. That way, Amethyst could be run in headless mode, which is exactly what we need for servers or bots.

**Comment by khionu** [comment]
I think that would require more discussion on the purpose of the `amethyst` crate. As is, I was under the assumption that it was meant to be a quick import for a game, rather than needing to specify all the individual crates.

**Comment by minecrawler** [comment]
Then you are questioning this whole PR?

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ it is attached to. ([#1282])
 * Add `icon` to `DisplayConfig` to set a window icon using a path to a file ([#1373])
 * Added setting to control gfx_device_gl logging level separately, and set it to Warn by default. ([#1404])
 * Add `loaded_icon` to `DisplayConfig` to set a window icon programatically ([#1405])
+* Added optional feature gates which will reduce compilation times when used. ([#1412])
```
Comment (torkleyy) [inline docs/CHANGELOG.md]: Can we change this to "which will reduce compilation time when disabled"?

**Comment by khionu** [comment]
No, I'm not questioning the whole PR. Adding the renderer as optional changes the scope of `amethyst` from "wrapper to help get going" to "the entry point that everything could use". If my understanding of the scope is wrong, then I would go further to say everything should be made optional and the crate refactored to ensure that that's possible. If my understanding is correct, the renderer would be a bit much to add to this PR.

Either way, this PR could be merged without your suggestion, and it could be moved to a follow up discussion/PR.

**Comment by torkleyy** [comment]
One request: can we plase make `gltf` optional? It's known to take very long to compile and I don't think it was enabled before #1411.

**Comment by fhaynes** [comment]
If we're going to do feature gates, please add renderer. Headless servers are a must and need to be in as soon as possible.

**Comment by fhaynes** [comment]
We can do that in a separate PR though

**Comment by jojolepro** [comment]
Ui and rendering: its more complex than the other ones, and they don't do anything if you are not using the bundles. Headless servers are already a thing this way.

I might try to add feature gates for those at a later time.

Also if I make gltf optional it won't get any testing from people using the cargo test commands.

**Comment by jojolepro** [comment]
Actually, if someone is up for it, feel free to open a new pull request after this one for input, ui and rendering. I'll continue to go through my todo list instead of going super deep in each element.

**Comment by torkleyy** [comment]
> Also if I make gltf optional it won't get any testing from people using the cargo test commands.

What about `cargo test --all`? Or `cargo test --all --all-features` (which should be used on CI)?

**Comment by jojolepro** [comment]
cargo test --all --all-features
works, but we have to update the pull request template so people don't forget. ;)

**Comment by torkleyy** [comment]
Accidentally? We want to test everything ;)

**Comment by jojolepro** [comment]
I assume most devs won't have installed the dependencies required for sdl testing.

**Comment by torkleyy** [comment]
I don't really see your point, sorry. If somebody wants to test everything, they just install the dependencies. However, testing is done by CI anyway. There is nothing `gltf` does that we couldn't test on CI.

And if that was the case, you could still do `cargo test --features gltf`.

**Comment by jojolepro** [comment]
we'll see if it works

**Comment by torkleyy** [comment]
Ah, nightly feature seems to be an issue :(

File: .concourse/tasks/linux-test.yml [hunk 2]
```diff
@@ -16,8 +16,9 @@ run:
   - |
     rustc --version
     export RUSTFLAGS="-D warnings"
-    cargo check --all --features sdl_controller,profiler
+    cargo check --all --features sdl_controller,profiler,gltf
```
Comment (LucioFranco) [inline .concourse/tasks/linux-test.yml]: `--all-features` here as well?

File: Cargo.toml [hunk 3]
```diff
@@ -19,6 +19,25 @@ license = "MIT/Apache-2.0"
...
+
+animation = [
+    "amethyst_animation"
+]
+audio = [
+    "amethyst_audio"
+]
+gltf = [
+    "amethyst_gltf",
+    "amethyst_animation"
```
Comment (LucioFranco) [inline Cargo.toml]: not totally positive but can we not reference the `animation` feature instead of the crate here?

File: docs/CHANGELOG.md [hunk 4]
```diff
@@ -66,6 +69,8 @@ it is attached to. ([#1282])
 [#1404]: https://github.com/amethyst/amethyst/pull/1404
 [#1408]: https://github.com/amethyst/amethyst/pull/1408
 [#1405]: https://github.com/amethyst/amethyst/pull/1405
+[#1411]: https://github.com/amethyst/amethyst/pull/1411
+[#1412]: https://github.com/amethyst/amethyst/pull/1412
```
Comment (torkleyy) [inline docs/CHANGELOG.md]: The entry for 1412 seems to be missing.

Comment (jojolepro) [reply docs/CHANGELOG.md]: * Added optional feature gates which will reduce compilation times when used. ([#1412])

File: Cargo.toml (hunk 3 above)
Comment (jojolepro) [reply Cargo.toml]: doesn't work

File: .concourse/tasks/linux-test.yml (hunk 2 above)
Comment (jojolepro) [reply .concourse/tasks/linux-test.yml]: would not work on nightly. Torkleyy made a change to support it in another PR.

File: book/src/appendices/c_feature_gates.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+Enabling the nightly feature adds a bit of debug information when running into runtime issues. To use it, you need to use the nightly rust compiler toolchain.
+
+Here is how to enable it:
+
+```
+cargo (build/test/run) --features nightly
+```
+
+The reason mentionned the most often for using it is to show exactly which resource is missing when a Resources::fetch() fails (or World::read_resource()).
```
Comment (azriel91) [inline book/src/appendices/c_feature_gates.md]: Maybe:

The most common use of this feature is to find out the type name of the resource that is missing, such as when a `Resources::fetch()` or `World::read_resource()` invocation fails.

Comment (jojolepro) [reply book/src/appendices/c_feature_gates.md]: This sounds way better than what I have, thanks!

**Comment by jojolepro** [comment]
bors retry

**Pull request merged**